Rover: BendyRuler and Dijkstra's path planning around obstacles

rmackay9 opened the pull request:
This PR adds two new methods of obstacle avoidance (BendyRuler and [Dijkstra's](https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm)) to the AC_Avoid and Rover.

I've marked this PR as a WIP but I think it is close-ish to being ready (once the "known issues" mentioned below are resolved).  I'd really like some feedback on the structure and anything else really to help me get it over the line.

[Tridge described BendyRuler in this video](https://www.youtube.com/watch?v=DZE_sl456dA) from the recent Developer's Unconference in Canberra but in short, it probes around the vehicle in 5deg increments and checks how far the probe point (or more accurately the line between the vehicle and the probe point) come to obstacles including the circular and polygon fences and barriers detected by the proximity sensor.

[Dijkstra's algorithm is descrbied to some degree here on Wikipedia](https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm) but in short does this:

- creates a list of safe nodes which are a few meters inside the polygon fence
- calculates the distances between all the nodes and puts these into "visibility graphs"
- iterates through the nodes (starting at the "source" and ending at the "destination") updating each nodes distance from the source.  After updating all the distances the shortest path is clear by iterating backwards from destination to source and making use of the "distance_from_idx" held for each node.

Both methods have also been tested on a real vehicle (my AION Robotics R1 rover).  Some videos from testing:

- [SITL video of Dijkstra's](https://www.youtube.com/watch?v=JKxQGt6XW60)
- [SITL video of BendyRuler](https://www.youtube.com/watch?v=KuJLxMEcPnU)

WickedShell commented on libraries/AP_Math/polygon.cpp:
```diff
@@ -109,3 +111,84 @@ template bool Polygon_outside<int32_t>(const Vector2l &P, const Vector2l *V, uns
+  determine if the polygon of N verticies defined by points V is
+  intersected by a line from point p1 to point p2
+ */
+bool Polygon_intersects(const Vector2f *V, unsigned N, const Vector2f &p1, const Vector2f &p2, Vector2f &intersection)
```
This should ideally be templated so we can leverage it against the Vector2l polygon's we have. (Quite useful for path generation for a survey grid for instance)

WickedShell commented on libraries/AP_Math/polygon.cpp:
```diff
@@ -109,3 +111,84 @@ template bool Polygon_outside<int32_t>(const Vector2l &P, const Vector2l *V, uns
+ */
+bool Polygon_intersects(const Vector2f *V, unsigned N, const Vector2f &p1, const Vector2f &p2, Vector2f &intersection)
+{
+    bool intersects = false;
```
You don't actually need `intersects` just set `intersect_dist_sq = FLT_MAX`, (or infinity I guess) you will always step info the if case for dist_sq, and then can just `return intersect_dist_sq < FLT_MAX;` at the end. Saves you a negation, logical or, and assignment per polygon line that intersects.

rmackay9 replied:
ok txs.  intersects boolean removed

WickedShell commented on libraries/AP_Math/vector2.cpp:
```diff
@@ -126,11 +131,11 @@ bool Vector2<T>::operator !=(const Vector2<T> &v) const
 float Vector2<T>::angle(const Vector2<T> &v2) const
 {
-    const float len = this->length() * v2.length();
+    float len = this->length() * v2.length();
```
Why are you removing const?

rmackay9 replied:
I suspect this was a merge conflict as I pulled in some commits from the OBC branch.  restored!

WickedShell commented on libraries/AP_Math/vector2.cpp:
```diff
@@ -126,11 +131,11 @@ bool Vector2<T>::operator !=(const Vector2<T> &v) const
         return 0.0f;
     }
-    const float cosv = ((*this)*v2) / len;
+    float cosv = ((*this)*v2) / len;
```
Why are you removing const?

rmackay9 replied:
restored!

WickedShell commented on libraries/AP_Math/vector2.cpp:
```diff
@@ -237,8 +248,175 @@ bool Vector2<T>::circle_segment_intersection(const Vector2<T>& seg_start, const
+    return v * (*this * v)/(v*v);
+}
+
+// given a position p1 and a velocity v1 produce a vector
```
Description doesn't match the function arguments. Leaves you guessing that `p1` is `pos_delta`

rmackay9 replied:
I'm pretty sure this was an existing issue with the comments before the function was moved but I'll fix it anyway, txs.

WickedShell replied:
Quite possibly, I didn't catch until later that a lot of stuff in this file was moving it out of the header or not, which meant I had started reviewing old/pre existing code/made it harder to see what was actually a change.

rmackay9 replied:
sorry about that.. I took a short-cut when merging from the OBC branch because I just took the complete file instead of trying to find all the patches that made up the change.

WickedShell replied:
Ah. That's really bad actually. Means this needs a much more detailed review to catch things that actually changed between master and this file (IE the problem introduced back into the polygon code)

rmackay9 replied:
ok, I'll go back and split the changes into two commits - one for just the code moving, another for he actual change.

WickedShell commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,343 @@
+        _prx_loc_count = count;
+    }
+
+    // even on failure to retrieve latest locations (which can happen due to a semaphore conflict)
```
How does this ever fail due to a semaphore conflict? It's a blocking take on the semaphore, so as far as I can tell this comment is wrong/misleading. The only time `copy_locations` fails is if we give it a null pointer, which 

A: we shouldn't be, passing by reference would be safer
B: if we did somehow then we shouldn't be continuing anyways

rmackay9 replied:
ok, i've changed it to return false if it can't copy_locations.  also removed the unnecessary "count" local variable.

rmackay9 commented:
@WickedShell, thanks very much for the review!

Some other questions I have for those interested:

1. do people think it would probably be better to make a proper frontend/backend split?  I suspect it would be.  I haven't yet and perhaps that's because:
     - the interfaces are slightly different (BendyRuler requires a velocity vector, Dijkstra's doesn't)
     - we may eventually want the user to be able to run both algorithsm (Dijkstra's first, Bendy later) although this hasn't been tested and may not work.
2. Dijkstra's needs arrays (for the visibility graph, nodes and return path) that must be larger if there are more polygon points.  The relationship is not exactly 1-to-1, it's more related to how many polygon points can see each other (i.e. a big circle is the worst case).  To get the arrays to be large enough without wasting a ton of memory I'm wondering if it would be better to:
  a. have a parameter that somehow controls the size of the array.
  b. create the arrays only after we know how many polygon points we have.
  In either case I'll probably need to add some error reporting if the arrays fill up.

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,343 @@
+bool AP_OABendyRuler::update(const Location& current_loc, const Location& destination, const Vector2f &ground_speed_vec, Location &origin_new, Location &destination_new)
+{
+    // exit immediately if neither circular nor polygon fence is enabled
+    const AC_Fence *fence_ptr = AC_Fence::get_singleton();
```
thanks.  I'm going to remove this check completely - it was added early on before we added the support for the proximity sensors.

WickedShell replied:
AC_Fence isn't required to be around, so you do need the nullptr check.

rmackay9 replied:
I've also removed the use of fence-ptr's margin in this update() method so it will only be used down in the calc_margin_from_xxxx_fence calls which do include the nullptr check.

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,343 @@
+    // and right. For each direction check if vehicle would avoid all obstacles
+    float best_bearing = bearing_to_dest;
+    bool have_best_bearing = false;
+    float best_margin = OA_VERY_BAD_MARGIN;
```
sure replaced VERY_BAD_MARGIN with -FLT_MAX and very good margin with +FLT_MAX.

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,343 @@
+        }
+        // bearing that we are probing
+        const float bearing_test = wrap_180(bearing_to_dest + bearing_delta);
+        const float bearing_change_deg = wrap_180(bearing_test - ground_course_deg);
```
yes, it's certainly possible that it might be noisy.  It's got a "ground_speed_vec.is_zero()" check further up but the EKF is unlikely to report an exactly zero speed.  It's more likely to be low.. so we could add some minimum speed in there instead.. maybe 0.2m/s?

rmackay9 replied:
I've replaced the is_zero with a OA_BENDYRULER_LOW_SPEED_SQUARED definition of 0.04f (ie. 0.2m/s)

rmackay9 replied:
This has definitely improved behaviour by the way.  It would sometimes become indecisive, i was unsure of the cause but I think this was it.

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,343 @@
+    float best_margin = OA_VERY_BAD_MARGIN;
+    float best_margin_bearing = best_bearing;
+
+    for (uint16_t i = 0; i < 360 / OA_BEARING_INC; i++) {
```
thanks, static assert added along with an extra pair of brackets in the for loop

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,343 @@
+    }
+
+    // debug
+    static uint32_t debug_ms = 0;
```
oops, that was some debug left over, zapped.

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,343 @@
+    const float fence_radius = fence_ptr->get_radius();
+
+    // margin is fence radius minus the longer of start or end distance
+    margin = fence_radius - MAX(start_dist, end_dist);
```
I think this is correct because of the minus ahead of the MAX.  So what we have is a line segment with start_dist being one end's distance from home (which is also the center of the circular fence) and the end_dist being the other end's distance from home.  the margin gets smaller as either of these distances approaches the fence_radius.

WickedShell replied:
Argh, to tired when reviewing, you're right.

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,343 @@
+// calculate minimum distance between a path and any obstacle
+float AP_OABendyRuler::calc_avoidance_margin(const Location &start, const Location &end)
+{
+    float circular_fence_margin = OA_VERY_GOOD_MARGIN;
```
I've replaced OA_VERY_GOOD_MARGIN with FLT_MAX.

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,343 @@
+float AP_OABendyRuler::calc_avoidance_margin(const Location &start, const Location &end)
+{
+    float circular_fence_margin = OA_VERY_GOOD_MARGIN;
+    calc_margin_from_circular_fence(start, end, circular_fence_margin);
```
I think what is below might be the alternative.  It seems a bit pointless in that it's more lines to do the same thing.. but I see your point about the return not being used.  It seemed bad (when writing the function) to simply fail to update the passed in argument without returning a boolean.  Anyway, I'll change to what we have below.

> float circular_fence_margin;
> if (!calc_margin_from_circular_fence(start, end, circular_fence_margin)) {
>     circular_fence_margin = FLT_MAX;
> }

rmackay9 commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -0,0 +1,193 @@
+        _thread_created = true;
+    }
+
+    uint32_t now = AP_HAL::millis();
```
const added, txs

rmackay9 commented on libraries/AP_Proximity/AP_Proximity.cpp:
```diff
@@ -378,6 +379,78 @@ const Vector2f* AP_Proximity::get_boundary_points(uint16_t& num_points) const
+
+    // copy locations into caller's buffer
+    num_copied = MIN(_location_count, buff_size);
+    for (uint16_t i=0; i<num_copied; i++) {
```
sure, replaced with a memcpy..

rmackay9 commented on libraries/AP_Proximity/AP_Proximity.cpp:
```diff
@@ -378,6 +379,78 @@ const Vector2f* AP_Proximity::get_boundary_points(uint16_t& num_points) const
+    WITH_SEMAPHORE(_rsem);
+
+    // get number of objects, return early if none
+    uint8_t num_objects = get_object_count();
```
const added

rmackay9 commented on libraries/AP_Proximity/AP_Proximity.cpp:
```diff
@@ -378,6 +379,78 @@ const Vector2f* AP_Proximity::get_boundary_points(uint16_t& num_points) const
+    // convert offset from vehicle position to earth frame location
+    const uint32_t now_ms = AP_HAL::millis();
+    _location_count = 0;
+    for (uint8_t i=0; i<num_objects; i++) {
```
thanks for this.  This looks like a real bug.  I think _locations[] probably needs to be up to PROXIMITY_SECTORS_MAX instead of PROXIMITY_MAX_DIRECTION and this leads us to move it to the AP_ProximityBackend.. which is probably where it belongs.

rmackay9 commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -0,0 +1,193 @@
+    avoidance_request.origin = origin;
+    avoidance_request.destination = destination;
+    avoidance_request.ground_speed_vec = AP::ahrs().groundspeed_vector();
+    if (!AP::ahrs().airspeed_estimate(&avoidance_request.airspeed)) {
```
yes, that's right.  We don't normally add stuff that is not used though so I have removed it in a follow up commit, txs.

rmackay9 commented on libraries/AP_Math/polygon.cpp:
```diff
@@ -43,10 +45,10 @@ bool Polygon_outside(const Vector2<T> &P, const Vector2<T> *V, unsigned n)
-        const T dx2 = V[j].x - V[i].x;
-        const T dy1 = P.y - V[i].y;
-        const T dy2 = V[j].y - V[i].y;
+        const int32_t dx1 = P.x - V[i].x;
```
fixed the reversion in a follow up commit.

rmackay9 commented:
I've reworked many of the commits in this PR to address the review issues from @WickedShell (thanks!) and also to resolve some bugs found and ToDo's that were listed above (now deleted):

1. Split the AP_Math commit into many smaller commits to make reviewing easier.  Also fixed some regressions that @WickedShell had found.
2. Search algorithms don't consume memory when disabled (Dijkstra's in particular eats just under 2.5k when enabled).
3. Dijkstra's detects polygon fence changes and reruns shortest-path calculations
4. Rebased on master to resolve conflicts with recent fence changes

I think this PR is in much better shape now and (in my mind) close to being ready for master so any further reviews are very much appreciated!

rmackay9 commented:
I've rebased this to resolve some minor merge conflicts from other PRs.

WickedShell commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    // fail if fence is too large
+    if (num_points > OA_DIJKSTRA_POLYGON_FENCE_PTS) {
```
Pretty sure this need to be `>=` because you copy the first point to be the last one, which means you can only accept 1 fewer point into `_polyfence_pts`

rmackay9 replied:
ah, thanks.  with Peter's recent changes we don't need to close the polygon anymore so I'll remove that bit.

WickedShell commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+        // find points before and after current point (relative to current point)
+        const uint8_t before_idx = (i == 0) ? num_points-1 : i-1;
+        Vector2f before_pt = boundary[before_idx] - boundary[i];
+        Vector2f after_pt = boundary[i+1] - boundary[i];
```
On the last loop doesn't this index the boundary at `boundary[num_points]` which should be after the end of the array we were supposed to be looking into?

rmackay9 replied:
yes, that's a straight up bug which again crept in when the fence boundary definition changed recently.  Anyway, txs for finding, fixed! (I think)

WickedShell commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    // clear existing fence
+    _polyfence_numpoints = 0;
```
Just set this to `num_points + 1`, then you can skip the work on lines 181/186. Line 185 just becomes _polyfence_pts[num_points] = ...`

rmackay9 replied:
great, thanks.  It's _polyfence_numpoints = num_points now

WickedShell commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+    // calculate distance from each polygon fence point to all other points
+    for (uint8_t i=0; i<_polyfence_numpoints-1; i++) {
+        for (uint8_t j=i+1; j<_polyfence_numpoints-1; j++) {
+            const Vector2f &start1 = _polyfence_pts[i];
```
Take this reference outside the inner loop so you can reuse it, without having to recalculate it all the time.

rmackay9 replied:
nice, thanks.

WickedShell commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+            if (!Polygon_intersects(boundary, num_points, start1, end1, intersection)) {
+                // line segment does not intersect with original fence so add to visgraph
+                _polyfence_visgraph.add_item({AP_OAVisGraph::OATYPE_FENCE_POINT, i},
+                                              {AP_OAVisGraph::OATYPE_FENCE_POINT, j},
```
Whitespace misalignment here is unfortunate.

rmackay9 replied:
whitespace fixed, txs

WickedShell commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+                // line segment does not intersect with original fence so add to visgraph
+                _polyfence_visgraph.add_item({AP_OAVisGraph::OATYPE_FENCE_POINT, i},
+                                              {AP_OAVisGraph::OATYPE_FENCE_POINT, j},
+                                              safe_sqrt(sq(_polyfence_pts[i].x - _polyfence_pts[j].x) + sq(_polyfence_pts[i].y - _polyfence_pts[j].y)));
```
Not that important, but could you do all your math with distance squared so you can save the square root time?

rmackay9 replied:
I had the same thought originally (i.e. use squares instead of actual distances) but we need to add distances and compare them later and this can lead to incorrect results if we use squared numbers.

i.e. is the length of dist1 + dist2 > dist3 + dist4?
dist1 = 6
dist2 = 6
dist3 = 1
dist4 = 10

dist1+dist2 is longer when using regular distances:
dist1+dist2 = 12
dist3+dist4 = 11

but sq(dist3) + sq(dist4) appears longer:
sq(dist1)+sq(dist2) = 72
sq(dist3)+sq(dist4) = 101

WickedShell commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+    update_visgraph(_destination_visgraph, {AP_OAVisGraph::OATYPE_DESTINATION, 0}, destination_NE);
+
+    // clear short_path_data array
+    _short_path_data_numpoints = 0;
```
This is a very verbose way to set the first and second points with the total number of points set to 2.

rmackay9 replied:
sure, shortened to 3 lines..

WickedShell commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+    // skip last fence point because it is the same as the first
+    for (uint8_t i=0; i<_polyfence_numpoints-1; i++) {
+        _short_path_data[_short_path_data_numpoints] = {{AP_OAVisGraph::OATYPE_FENCE_POINT, i}, false, OA_DIJKSTRA_POLYGON_VISGRAPH_INFINITY_CM, OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX};
+        _short_path_data_numpoints++;
```
You don't need to increment in the loop, you know exactly how many points you just added.

rmackay9 replied:
I'll move the increment into the line above.. from a processing time point of view I don't think it matters because we either need to add (i+_short_path_data_numpoints) or increment _short_path_data_numpoints.

WickedShell replied:
You don't need an increment. You can directly add `_polyfence_numpoints-1` outside of the for loop. Which saves you from doing it each iteration.

rmackay9 replied:
well, we'd either need an addition or an increment.. so not sure there's much of a difference..

WickedShell replied:
In one case you do it for each point, so you do the increment `n` times. The other is a single add, which is faster then doing it several times.

rmackay9 replied:
not quite.. what I mean is that within the for loop I'd need to have an addition like i+2 so there's an addition happening anyway.

> _short_path_data[i+2] = {{AP_OAVisGraph::OATYPE_FENCE_POINT, i}, false, OA_DIJKSTRA_POLYGON_VISGRAPH_INFINITY_CM, OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX};

WickedShell commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+            return false;
+        }
+        _path_numpoints = 0;
+        bool complete = false;
```
You don't need the complete variable, it actually makes it more complicated (and slower). Just make this a while (true) and break once you get to the complete case.

rmackay9 replied:
ok, I feel odd about having a, "while (true)" in there but I'll change it txs..

WickedShell commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+}
+
+// return number of points in shortest path
+uint8_t AP_OADijkstra::shortest_path_numpoints()
```
Do we need this helper? It's a one line helper, that isn't used anywhere else and impairs the readability in the one spot it's used (where it's used differently/as part of a debug interface).

rmackay9 replied:
No we don't.  I originally added it when I thought we needed a semaphore around the final path.  We may need that eventually but for the moment only the backend ever accesses the path so removing. txs

WickedShell commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+}
+
+// print shortest path for debug purposes
+void AP_OADijkstra::print_shortest_path()
```
Add debug guards?

rmackay9 replied:
txs, i've removed both print_ methods.

WickedShell commented on libraries/AC_Avoidance/AP_OAVisGraph.h:
```diff
@@ -0,0 +1,69 @@
+    const VisGraphItem& get_item(uint8_t index_num) const { return _items[index_num]; }
+
+    // allow accessing graph as an array, 0 indexed
+    VisGraphItem& operator[](uint8_t i) { return _items[i]; }
```
If you failed to init and allocate the array this will be a nullpointer dereference.

rmackay9 replied:
we never used the unconst'd version of operator[] so I've removed it (thus reducing danger)

WickedShell commented on libraries/AC_Avoidance/AP_OAVisGraph.h:
```diff
@@ -0,0 +1,69 @@
+    bool add_item(const OAItemID &id1, const OAItemID &id2, float distance_cm);
+
+    // get item by index number
+    const VisGraphItem& get_item(uint8_t index_num) const { return _items[index_num]; }
```
If you failed to init and allocate the array this will be a nullpointer dereference.

rmackay9 replied:
we never actually use get_item so I've removed it. txs

WickedShell commented on libraries/AC_Avoidance/AP_OAVisGraph.h:
```diff
@@ -0,0 +1,69 @@
+
+    // allow accessing graph as an array, 0 indexed
+    VisGraphItem& operator[](uint8_t i) { return _items[i]; }
+    const VisGraphItem& operator[](uint8_t i) const { return _items[i]; }
```
If you failed to init and allocate the array this will be a nullpointer dereference.

rmackay9 replied:
the VisGraph::operator[] it's a little dangerous perhaps but at the moment we only use it within loops that use the num_items() function so they will never try to access invalid memory.  I'll add a note to warn that there are no protections.

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+extern const AP_HAL::HAL &hal;
+
+#define OA_DIJKSTRA_POLYGON_VISGRAPH_PTS         (OA_DIJKSTRA_POLYGON_FENCE_PTS * OA_DIJKSTRA_POLYGON_FENCE_PTS / 2)
+#define OA_DIJKSTRA_POLYGON_VISGRAPH_INFINITY_CM 99999999.0f
```
sure, replaced big constant with FLT_MAX

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+
+#define OA_DIJKSTRA_POLYGON_VISGRAPH_PTS         (OA_DIJKSTRA_POLYGON_FENCE_PTS * OA_DIJKSTRA_POLYGON_FENCE_PTS / 2)
+#define OA_DIJKSTRA_POLYGON_VISGRAPH_INFINITY_CM 99999999.0f
+#define OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX 255     // index use to indicate we do not have a tentative short path for a node
```
I guess you mean add a static_assert that **OA_DIJKSTRA_POLYGON_FENCE_PTS** < OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX.

Added, thanks!

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+}
+
+// print out all node data for debug purposes
+void AP_OADijkstra::print_all_nodes()
```
txs, i've removed both print_ methods.

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+        Vector2f origin_pos;
+        if ((_path_idx_returned > 0) && get_shortest_path_point(_path_idx_returned-1, origin_pos)) {
+            // convert offset from ekf origin to Location
+            Location temp_loc(Vector3f(origin_pos.x, origin_pos.y, 0.0f));
```
There is protection but it's quite a ways away in the code.  To be specific the AC_Fence::get_boundary_points() method returns the fence points as offsets from the EKF origin which will only succeed if we have an EKF origin.  The exact check is within the AC_Fence::load_polygon_from_eeprom() method.

rmackay9 replied:
Happy to add another check though if you think that would be best..

WickedShell replied:
It's getting deep in the call hierarchy, if we have protection as long as people outside the library can't get to this it's probably fine. It gets to be hard to figure out how to bubble these up and out. I don't want to sink this in paranoid checks if we have protection, it just wasn't obvious to me at the time if there was any. Your call either way though.

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+bool AP_OADijkstra::find_node_from_id(const AP_OAVisGraph::OAItemID &id, node_index &node_idx) const
+{
+    // source node is always the first node
+    if ((id.id_type == AP_OAVisGraph::OATYPE_SOURCE) && (_short_path_data_numpoints > 0)) {
```
sure, thanks.  replaced ifs with switch statement.

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+bool AP_OADijkstra::find_closest_node_idx(node_index &node_idx) const
+{
+    uint16_t lowest_idx = 0;
+    float lowest_dist = 0.0f;
```
not sure what is meant by the "storing as float" and "cast back down later".. for the indices it's all ints.  For the distances it's floats.. but I don't see any casting back to int..

For the 2nd comment, yes we can use your favourite FLT_MAX and remove the boolean.  done, txs!

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.h:
```diff
@@ -0,0 +1,115 @@
+        bool visited;                   // true if all this node's neighbour's distances have been updated
+        float distance_cm;              // distance from source (number is tentative until this node is the current node and/or visited = true)
+        node_index distance_from_idx;   // index into _short_path_data from where distance was updated (or 255 if not set)
+    } _short_path_data[OA_DIJKSTRA_POLYGON_SHORTPATH_PTS];
```
ok txs.  it makes it less readible but for 4bytes per entry that's quite valuable. excuse my ignorance but how about if "id" was placed below "distance_from_idx" would the 4 bytes be saved then too?

WickedShell replied:
Structs (unless you tag them with `PACKED` (which is actually a macro for compiler attributes)), have all their members aligned on word access. So the id is 1 byte, that's fine. But the 4 byte float won't take 3 bytes from the first word and 1 from the second as that's slower. But this meant you had 3 bytes of padding that weren't doing anything for you.

You could tag this PACKED and keep the current structure, but it would slow down the runtime performance. Rearranging can't always get as tight as packed can, but it doesn't sacrifice runtime performance.

rmackay9 replied:
ok, cool, txs

rmackay9 replied:
ah, actually the id is two bytes.. so i'll re-order it like this:

>     struct ShortPathNode {
>         AP_OAVisGraph::OAItemID id;     // unique id for node (combination of type and id number)
>         bool visited;                   // true if all this node's neighbour's distances have been updated
>         node_index distance_from_idx;   // index into _short_path_data from where distance was updated (or 255 if not set)
>         float distance_cm;              // distance from source (number is tentative until this node is the current node and/or visited = true)
>     } _short_path_data[OA_DIJKSTRA_POLYGON_SHORTPATH_PTS];

WickedShell commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+// returns true if successful and node_idx argument is updated
+bool AP_OADijkstra::find_closest_node_idx(node_index &node_idx) const
+{
+    uint16_t lowest_idx = 0;
```
You're still mixing types on this though. It's `uint16_t` but you are down casting to `node_index`. These should be consistent if you are going to use the typedefs like this.

rmackay9 replied:
yes, you're right.  I'll replace the uint16_t with node_index.  txs again

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.h:
```diff
@@ -0,0 +1,115 @@
+    // updates visibility graph for a given position which is an offset (in cm) from the ekf origin
+    // requires create_polygon_fence_with_margin to have been run
+    // returns true on success
+    bool update_visgraph(AP_OAVisGraph& visgraph, const AP_OAVisGraph::OAItemID& oaid, const Vector2f &position, bool add_extra_point = false, Vector2f extra_point = Vector2f(0,0));
```
added comment, thanks!

I'll also rename the "extra_point" argument to "extra_position" so we don't unnecessarily mix words for the same thing.

rmackay9 commented:
I've addressed the latest issues from @WickedShell.  Thanks very much for the review, really really appreciated!

rmackay9 commented:
I've pushed a small but important fix to bendy ruler.  It was causing the vehicle to refuse to fly directly south (:-)) which showed up in [this latest video I recorded earlier](https://www.youtube.com/watch?v=GAmNaDTzy3Q).

peterbarker commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+        test_loc.offset_bearing(bearing_test, lookahead_step1_dist);
+
+        // calculate margin from fence for this scenario
+        float margin = calc_avoidance_margin(current_loc, test_loc);
```
`const`

peterbarker commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+            best_margin = margin;
+        }
+        if (margin > _margin_max) {
+            // this bearing avoids obstacles out to the lookahead_step1_dist
```
Comment is incorrect.  We should probably also have sanity checks that relate your `_margin_max` to `lookahead_step1_dist`

rmackay9 replied:
Is the comment incorrect?  In any case, I've changed the following comment to, "// now check in there is a clear path in three directions towards the destination".
I've added sanity checking to the AP_OABendyRuler::set_config() call so we can be sure the _look_ahead distance is at least 1m and _margin_max is at least positive.  I've done something similar to AP_OADijkstra::set_fence_margin to ensure it gets a positive margin

peterbarker commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+            }
+
+            // perform second stage test in three directions looking for obstacles
+            const float test_bearings[] = { 0, 45, -45 };
```
Should be floating point constants.  And can lose the `=`.

rmackay9 replied:
changed to floating point constants and removed the "=". txs

peterbarker commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+                    destination_new = current_loc;
+                    destination_new.offset_bearing(bearing_test, distance_to_dest);
+                    _current_lookahead = MIN(_lookahead, _current_lookahead * 1.1f);
+                    return (i != 0 || j != 0);
```
The description on this function is

```
 // returns true and updates origin_new and destination_new if a best path has been found
```

Why is `i == 0 && j == 0` not a valid best path?

rmackay9 replied:
if i and j are zero then the best path is directly towards the destination so we might as well turn off object avoidance.  I've added a comment to clarify this, "// if the chosen direction is directly towards the destination turn off avoidance"

peterbarker commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+    return true;
+}
+
+// calculate minimum distance between a path and any obstacle
```
```suggestion
// calculate minimum distance between a segment and any obstacle
```

rmackay9 replied:
changed, "path" to "segment", txs

peterbarker commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+    }
+
+    // return smallest margin from any obstacle
+    return MIN(MIN(circular_fence_margin, polygon_fence_margin), proximity_margin);
```
This is going to get nasty :-)

Next time somethings added we probably want a `float smallest_margin = FLT_MAX` at the top of this function.  Each clause in here would stop negating the result of the method call and instead do the `MIN(...)` inside the body of the `if`.

rmackay9 replied:
ok, I guess we could change the functions so they don't return a boolean and instead return FLT_MAX if they can't calculate a margin for some reason (i.e. the fence doesn't exist).

peterbarker commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+{
+    // exit immediately if polygon fence is not enabled
+    const AC_Fence *fence_ptr = AC_Fence::get_singleton();
+    if (fence_ptr == nullptr) {
```
I can *see* that it's a pointer :-)

rmackay9 replied:
ok, renamed all local variables which had the _ptr appended to them

peterbarker commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+    }
+
+    // calculate start and end point's distance from home
+    const Location ahrs_home = AP::ahrs().get_home();
```
```suggestion
    const Location &ahrs_home = AP::ahrs().get_home();
```

rmackay9 replied:
txs for that.  By the way, we have a similar mistake in GCS_Common.cpp's GCS_MAVLINK::send_home_position.  I might create a small PR for this fix.

peterbarker commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+// calculate minimum distance between a path and any obstacle
+float AP_OABendyRuler::calc_avoidance_margin(const Location &start, const Location &end)
+{
+    float circular_fence_margin;
```
This should be the margin squared - that will save a *lot* of `sqrtf`s!

rmackay9 replied:
great suggestion!  .. but I've investigated returning all the margins as squared values but I can't see how we can do that because for each of the three cases, when calculating the margins we need to do some addition/subtraction and these aren't commutative when dealing with squared values.  I.e. sq(a+b) != sq(a)+sq(b).

anyway, I've managed to reduce the number of sqrt calls by one by reordering calc_margin_from_circular_fence a bit.

OXINARF commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -395,20 +401,18 @@ bool AP_OADijkstra::calc_shortest_path(const Location &origin, const Location &d
+    // check OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX is defined correct
+    static_assert(OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX > OA_DIJKSTRA_POLYGON_FENCE_PTS, "check OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX > OA_DIJKSTRA_POLYGON_FENCE_PTS");
 
     // add origin and destination (node_type, id, visited, distance, distance_from_idx) to short_path_data array
```
Comment needs to be changed to be according to new struct order.

rmackay9 replied:
sorry, I don't know what this means..

rmackay9 replied:
fixed, thanks!

OXINARF commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -395,20 +401,18 @@ bool AP_OADijkstra::calc_shortest_path(const Location &origin, const Location &d
+    _short_path_data[1] = {{AP_OAVisGraph::OATYPE_DESTINATION, 0}, false, OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX, FLT_MAX};
+    _short_path_data_numpoints = 2;
 
     // add fence points to short_path_data array (node_type, id, visited, distance, distance_from_idx)
```
Same.

rmackay9 replied:
fixed, thanks!

rmackay9 commented on APMrover2/AP_Arming.cpp:
```diff
@@ -166,3 +167,20 @@ bool AP_Arming_Rover::disarm(void)
+// check object avoidance has initialised correctly
+bool AP_Arming_Rover::oa_check(bool report)
+{
+    const char *fail_msg = nullptr;
```
ok changed AP_OAPathPlanner::pre_arm_check() so it accepts a buffer to write the error message into.

rmackay9 commented on APMrover2/AP_Arming.cpp:
```diff
@@ -166,3 +167,20 @@ bool AP_Arming_Rover::disarm(void)
+    }
+
+    // display failure
+    if (fail_msg == nullptr) {
```
At the moment at least, the pre_arm_check call always returns a string but I've incorporated the strlen just in case. txs

rmackay9 commented on APMrover2/mode_auto.cpp:
```diff
@@ -137,7 +137,7 @@ bool ModeAuto::get_desired_location(Location& destination) const
     switch (_submode) {
     case Auto_WP:
         if (g2.wp_nav.is_destination_valid()) {
-            destination = g2.wp_nav.get_destination();
```
I agree the name, "get_oa_destination()" is not fantastic.. this is actually not the "final" destination though, it's the intermediate destination() if there is one, and if there isn't an intermediate destination (i.e. no obstacles in the way to the destination) then it's the same as the regular "final" destination.

I think there's a bit more architectural cleanup around how we handle these waypoints but I'm not sure what that will be just yet so I might leave this as-is.

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+#include <AC_Fence/AC_Fence.h>
+#include <stdio.h>
+
+extern const AP_HAL::HAL &hal;
```
remove hal and the include of stdio.h.  They're convenient to have around for debugging but I'll remove them in any case, txs.

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+
+extern const AP_HAL::HAL &hal;
+
+const int16_t OA_BENDYRULER_BEARING_INC = 5;            // check every 5 degrees around vehicle
```
re the 2nd comment, "uint16_t", it's just going to get promoted to a uint16_t anyway when we do the (360 / this-constant) and we're not working with 8bit CPUs anymore anyway so I'm not sure it helps much so I might skip this one..

peterbarker replied:
You'll just get a small amount of compiler-time bounds checking.

If you omit the `=` sign and instantiate it like this:

```
const int8_t OA_BENDYRULER_BEARING_INC{5};
```

You get an even stronger check from the compiler.

It really doesn't matter here, but it's a good thing we can be doing in the code.

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+const float OA_BENDYRULER_LOOKAHEAD_STEP2_RATIO = 1.0f; // step2's lookahead length as a ratio of step1's lookahead length
+const float OA_BENDYRULER_LOOKAHEAD_STEP2_MIN = 2.0f;   // step2 checks at least this many meters past step1's location
+const float OA_BENDYRULER_LOOKAHEAD_PAST_DEST = 2.0f;   // lookahead length will be at least this many meters past the destination
+const float OA_BENDYRULER_LOW_SPEED_SQUARED = 0.04f;    // when ground course is below this speed squared, vehicle's heading will be used (0.04 = 0.2m/s)
```
changed definition to (0.2f*0.2f). txs

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+
+// run background task to find best path and update avoidance_results
+// returns true and updates origin_new and destination_new if a best path has been found
+bool AP_OABendyRuler::update(const Location& current_loc, const Location& destination, const Vector2f &ground_speed_vec, Location &origin_new, Location &destination_new)
```
ok, point taken... but in Copter and Rover we very commonly use "origin" when referring to the starting point of a journey to a waypoint.  I.e. the start of the segment.  We could perhaps move to "source" instead of "origin" but it's a relatively large renaming exercise that I don't think I'll take on as part of this PR.

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+    // check OA_BEARING_INC definition allows checking in all directions
+    static_assert(360 % OA_BENDYRULER_BEARING_INC == 0, "check 360 is a multiple of OA_BEARING_INC");
+
+    // search in 5 degree increments around the vehicle alternating left
```
by default it is still 5 degree increments, unless someone changes OA_BENDYRULER_BEARING_INC.  I'll change the comment to replace "5" with "OA_BENDYRULER_BEARING_INC".

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+        const float bearing_test = wrap_180(bearing_to_dest + bearing_delta);
+        const float bearing_change_deg = wrap_180(bearing_test - ground_course_deg);
+
+        // don't consider 180 degree turns, as we can't predict which way we will turn
```
txs.  I've changed the comment to, "// don't consider turns over 170 degrees, as we can't predict which way we will turn"

rmackay9 commented on libraries/AP_Math/vector2.cpp:
```diff
@@ -248,9 +248,63 @@ bool Vector2<T>::circle_segment_intersection(const Vector2<T>& seg_start, const
+// given a position pos_delta and a velocity v1 produce a vector
+// perpendicular to v1 maximising distance from p1
+template <typename T>
+Vector2<T> Vector2<T>::perpendicular(const Vector2<T> &pos_delta, const Vector2<T> &v1)
```
At the risk of passing the buck, the AP_Math changes are from the OBC branch so I'll ask @tridge if he can explain them.  I'll remove the unused functions.

rmackay9 commented on libraries/AP_Math/polygon.cpp:
```diff
@@ -129,10 +129,11 @@ template bool Polygon_complete<float>(const Vector2f *V, unsigned n);
   intersected by a line from point p1 to point p2
  */
-bool Polygon_intersects(const Vector2f *V, unsigned N, const Vector2f &p1, const Vector2f &p2)
+bool Polygon_intersects(const Vector2f *V, unsigned N, const Vector2f &p1, const Vector2f &p2, Vector2f &intersection)
```
ok, copied comment from .h to .cpp. txs

rmackay9 commented on libraries/AP_Proximity/AP_Proximity.cpp:
```diff
@@ -378,6 +379,78 @@ const Vector2f* AP_Proximity::get_boundary_points(uint16_t& num_points) const
+// copy location points around vehicle into a buffer owned by the caller
+// caller should provide the buff_size which is the maximum number of locations the buffer can hold (normally PROXIMITY_MAX_DIRECTION)
+// num_copied is updated with the number of locations copied into the buffer
+// returns true on success, false on failure (should only happen if there is a semaphore conflict)
```
yup, ok.  fixed comment on AP_Proximity_Backends::copy_locations() to clarify that it only fails when buff is nullptr.

rmackay9 commented on libraries/AP_Proximity/AP_Proximity.h:
```diff
@@ -99,6 +108,16 @@ class AP_Proximity
     const Vector2f* get_boundary_points(uint16_t& num_points) const;
 
+    // get Location points around vehicle for use by avoidance in earth-frame
+    //   returns nullptr and sets num_points to zero if no boundary can be returned
```
I've apparently updated the comments so it doesn't say "boundary" anymore but instead says "locations.

rmackay9 commented on libraries/AC_Fence/AC_Fence.h:
```diff
@@ -99,7 +99,11 @@ class AC_Fence
     ///
 
+    /// returns true if polygon fence is valid (i.e. has at least 3 sides)
+    bool is_polygon_valid() const { return _boundary_valid; }
```
this sounds important but I'm not sure I totally understand.  I think the point is that the ground stations may be pushing a polygon fence to the vehicle and call this "is_polygon_valid" while that is happening.  So perhaps perhaps we should add a check that, "_boundary_num_points != _total" like we have in AC_Fence::polygon_fence_is_breached?

rmackay9 commented on libraries/AC_Fence/AC_Fence.h:
```diff
@@ -112,6 +112,9 @@ class AC_Fence
     void handle_msg(GCS_MAVLINK &link, mavlink_message_t* msg);
 
+    /// return system time of last update to the boundary (allows external detection of boundary changes)
+    uint32_t get_boundary_update_ms() const { return _boundary_update_ms; }
```
I could add this comment, "/// this time applies to the boundary loaded into memory and may not apply to the boundary in eeprom" although I wonder if any future will understand it enough to make it useful in their developments..

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+        float bearing_delta = i * OA_BENDYRULER_BEARING_INC * 0.5f;
+        if (i & 1) {
+            // we alternate between trying to the left and right of the target
+            bearing_delta = -bearing_delta;
```
txs for finding this.  I can't imagine this is intentional so I've changed it to be two loops instead of one.  So the external loop will be from 0 to 170, the internal loop will be 0 to 1.  So it will look like this:

>for (uint8_t i = 0; i <= (170 / OA_BENDYRULER_BEARING_INC); i++) {
>        for (uint8_t bdir = 0; bdir <= 1; bdir++) {
>            // skip duplicate check of bearing straight towards destination
>            if ((i==0) && (bdir > 0)) {
>                continue;
>            }

rmackay9 commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,294 @@
+    destination_new = current_loc;
+    destination_new.offset_bearing(chosen_bearing, distance_to_dest);
+
+    return true;
```
sorry if I'm being dense but when you say, "new destination might actually be the same...", the same as what?  Maybe you mean that in the section above, "if (have_best_bearing", the "best_bearing" and the "best_margin_bearing" could be the same...? sorry, i'm missing the point some how..

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+            _polyfence_pts[i] = boundary[i] - intermediate_pt;
+            if (Polygon_outside(_polyfence_pts[i], boundary, num_points)) {
+                // could not find a point on either side that was within the fence so fail
+                // this can happen if fence lines are closer than margin_m
```
comment fixed, txs!

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+                // line segment does not intersect with original fence so add to visgraph
+                _polyfence_visgraph.add_item({AP_OAVisGraph::OATYPE_FENCE_POINT, i},
+                                              {AP_OAVisGraph::OATYPE_FENCE_POINT, j},
+                                              safe_sqrt(sq(_polyfence_pts[i].x - _polyfence_pts[j].x) + sq(_polyfence_pts[i].y - _polyfence_pts[j].y)));
```
Don't think so.. I think "norm(_polyfence_pts[i], _polyfence_pts[j])" would be sqrt(sq(_polyfence_pts[i]) + sq(_polyfence_pts[j]) ... I think it could be the same as norm(_polyfence_pts[i] - _polyfence_pts[j]) which would be the same as (_polyfence_pts[i] - _polyfence_pts[j]).length().  I think perhaps I was looking to avoid the creation of intermediate Vector2f objects but this function is run so rarely it's probably doesn't matter if it's slightly less efficient.

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+        Vector2f intersection;
+        if (!Polygon_intersects(boundary, num_points, position, _polyfence_pts[i], intersection)) {
+            // line segment does not intersect with original fence so add to visgraph
+            visgraph.add_item(oaid, {AP_OAVisGraph::OATYPE_FENCE_POINT, i}, safe_sqrt(sq(position.x - _polyfence_pts[i].x) + sq(position.y - _polyfence_pts[i].y)));
```
.. changed to (position - _polyfence_pts[i]).length()

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+    if (add_extra_point) {
+        Vector2f intersection;
+        if (!Polygon_intersects(boundary, num_points, position, extra_point, intersection)) {
+            visgraph.add_item(oaid, {AP_OAVisGraph::OATYPE_DESTINATION, 0}, safe_sqrt(sq(position.x - extra_point.x) + sq(position.y - extra_point.y)));
```
changed to ((position - extrapoint).length())

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.h:
```diff
@@ -0,0 +1,115 @@
+    // shortest path state variables
+    bool _polyfence_with_margin_ok = false;
+    bool _polyfence_visgraph_ok = false;
+    bool _shortest_path_ok = false;
```
txs, removed initialisations

rmackay9 commented on libraries/AP_Math/polygon.cpp:
```diff
@@ -148,21 +149,29 @@ bool Polygon_intersects(const Vector2f *V, unsigned N, const Vector2f &p1, const
 /*
   return the closest distance that a line from p1 to p2 comes to an
   edge of closed polygon V, defined by N points
+  negative numbers indicate the lines cross with the negative size being the distance from p2 to the intersection point
```
It's the intersection point closest to p1.  I'll improve the comment to be, "negative numbers indicate the lines cross with the negative size being the distance from p2 to the intersection point closest to p1"

rmackay9 commented on libraries/AC_Avoidance/AP_OADijkstra.cpp:
```diff
@@ -0,0 +1,550 @@
+
+    // extract path starting from destination
+    bool success = false;
+    {
```
Ah, ok. block removed.  This was a hangover from when I thought I needed a semaphore to pass the path back to AR_WPNav.  In this final version the complete path to the destination is not exposed to AR_WPNav, instead AP_OAPathPlanning provides only the next destination.

rmackay9 commented:
@OXINARF,

Thanks very much for the review, really great stuff.  I think I've addressed most of the issues and replied with a couple of questions for those I wasn't sure about.

I totally agree that the back and forth between using Locations (i.e. lat/lon) and Vector2f (i.e. offsets-from-ekf-origin) is not great and I plan to consolidate on Vector2f in a follow-up PR (issue created here: https://github.com/ArduPilot/ardupilot/issues/11534).  At the risk of passing the buck, the use of Location comes mostly from the BendyRuler code which comes from the OBC branch which was built for Plane.  I refactored it but didn't want to make this fundamental change immediately.  The proximity changes (mostly from @magicrub) will be changed to Vector2f as well.

I've added some logging of the final destination and the intermediate destination of the OAPathPlanning class.  It's possible we should add more logging but let's see what we need once testers start trying it.  At least we will be able to see that it's running and how often it's updating.

I will ask @tridge about the Vector2f changes.  TBH, I mostly just copied these from the OBC branch so I don't totally understand the changes.

peterbarker commented:
On Fri, 7 Jun 2019, Randy Mackay wrote:

> +        // don't consider 180 degree turns, as we can't predict which way we will turn
> 
> txs. I've changed the comment to, "// don't consider turns over 170 degrees, as we can't predict which way we will turn"

I do this myself - including the constant as part of the comment - but it 
means the comments *always* get out of date with the code :-)

peterbarker commented on libraries/AC_Avoidance/AP_OABendyRuler.cpp:
```diff
@@ -0,0 +1,286 @@
+
+// calculate minimum distance between a path and the circular fence (centered on home)
+// on success returns true and updates margin
+bool AP_OABendyRuler::calc_margin_from_circular_fence(const Location &start, const Location &end, float &margin)
```
Does this need to check if we're actually inside the circlular fence?  Similarly for the other two cases.

peterbarker replied:
.... also, what happens if `start` or `end` are *signficantly* outside the circle?

rmackay9 replied:
@peterbarker, I think you're right that we should check if we are inside the fence so I've created an issue:https://github.com/ArduPilot/ardupilot/issues/11543.  So if the vehicle goes outside the circular or Polygon fence, it will then it will not get back inside the fence, instead it will actively drive around avoiding the fence (from the outside).  This can be reproduced in SITL.

rmackay9 commented:
@peterbarker, re the 170deg comments, I've reworked the way it calculates the angles (because of an issue brought up by Francisco) and so the comment is completely gone now..

rmackay9 commented:
I've squashed some commits together and rebased on master.

I'll push to master once travis says it's OK.
